packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,359 @@
+        # then s.path equals "/foo/bar/".
+        s.wfile.write("<p>You accessed path: %s</p>" % s.path)
+        s.wfile.write("</body></html>")
+
+    def do_POST(s):
+        """Respond to a GET request."""
+        pass
+
+
+class HTTPChannel:
+    """A wrapper for an RESTful API with an internal HTTPServer."""
+
+    def __init__(self,
+                 # address: Address,
+                 api_spec: str,
+                 host: str,
+                 port: int,
+                 excluded_protocols: Optional[Set[PublicId]] = None,
+                 ):
+        """
+        Initialize a channel.
+
+        :param address: the address
+        """
+        # self.address = address
+        self.host = host
+        self.port = port
+        # self.uri_base = f"http://{host}:{port}"
+        self.in_queue = None  # type: Optional[asyncio.Queue]
+        self.loop = None  # type: Optional[asyncio.AbstractEventLoop]
+        self.excluded_protocols = excluded_protocols
+        self.thread = Thread(target=self.receiving_loop)
+        self.lock = threading.Lock()
+        self.stopped = True
+        # Initializing the internal HTTP server
+        try:
+            self.server = HTTPServer((self.host, self.port), HTTPHandler)
+        except OSError:
+            logger.error(f"{host}:{port} is already in use, please try another Socket.")
+        # Process API specifications into Envelopes to be sent to InBox
+        self._process_api(api_spec)
+        logger.info("Initialized the HTTP Server.")
+
+    def _process_api(self, api_spec):
+        """Process the api_spec.
+
+        Decode all paths to get envelops each, and put it in the agent's inbox.
+        """
+        try:
+            paths = api_spec['paths']
+        except KeyError:
+            logger.error("Key 'paths' not found in API spec.")
+
+        for path_name, path_value in paths.items():
+            try:
+                envelope = _decode_path(path_name, path_value)
+                assert self.in_queue is not None, "Input queue not initialized."
+                assert self._loop is not None, "Loop not initialized."
+                # Ok to put these Envelopes in_queue during HTTPChannel initialization?
+                asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self._loop)
+            except ValueError:
+                logger.error(f"Bad formatted path: {path_name}")
+            except Exception as e:
+                logger.error(f"Error when processing a path. Message: {str(e)}")
+
+    def _decode_path(self, path_name: str, path_value: dict, separator: str = '/'):
+        """Path --> Envelope
+
+        Convert specified API path name and value into in_queue'd Envelope.
+        Envelope's msg consists of the response options for each REST req in path.
+        """
+        path_split = path_name.split(separator)
+
+        to = path_split[1].strip()  # to: name of Skill (petstoresim)?
+        sender = 'TBD'  # hardcoded for now
```

Comment:
One way would be to create a uuid for each request. Then the sender is the uuid. You send the envelope into the agent, whilst blocking on the response to the request. Once a matching envelope with the same uuid as received comes back from within the agent (or a timeout is hit) your serve the response.